### PR TITLE
Use unexpectedErrorCode for unexpected errors that we should alert on…

### PR DESCRIPTION
**Pull request by yuki-stripe**
… in STPPaymentHandler

## Summary
Now that STPPaymentHandler sends start and finish analytics (https://github.com/stripe/stripe-ios/pull/3484) and sends loggableUserInfo in those analytics (https://github.com/stripe/stripe-ios/pull/3486), this PR:
1. _Finishes_ the confirmation with an error instead of hanging, in certain cases
2. Corrects errors that were using the wrong/inaccurate error code
3. Adds an "unexpectedErrorCode" case for unexpected errors that should never happen and indicate a problem with the SDK/API that we should alert on

## Testing
Relying on existing unit tests 

## Changelog
Not user facing.

### CODE REVIEW

File: StripePayments/StripePayments/Source/PaymentHandler/STPPaymentHandler.swift [hunk 1]
```diff
@@ -927,7 +937,7 @@ public class STPPaymentHandler: NSObject {
             action.complete(
                 with: STPPaymentHandlerActionStatus.failed,
                 error: _error(
-                    for: .intentStatusErrorCode,
+                    for: .unexpectedErrorCode,
```
[the remark this answers is not in the document]
Comment (yuki-stripe) [reply StripePayments/StripePayments/Source/PaymentHandler/STPPaymentHandler.swift:940]: I'll update that comment. I think "unknown status" is better classified as an unexpected error; it really means something is terribly wrong.

File: StripePayments/StripePayments/Source/PaymentHandler/STPPaymentHandler.swift [hunk 2]
```diff
@@ -998,45 +1008,51 @@ public class STPPaymentHandler: NSObject {
...
     }
 
-            func _handleAuthenticationForCurrentAction() {
-        guard let currentAction = currentAction,
+    func _handleAuthenticationForCurrentAction() {
+        guard let currentAction,
             let authenticationAction = currentAction.nextAction()
         else {
+            stpAssertionFailure("Missing current action or next action!")
+            let errorAnalytic = ErrorAnalytic(event: .unexpectedPaymentHandlerError, error: InternalError.invalidState, additionalNonPIIParams: ["error_message": "Missing current action or next action!"])
```
[the remark this answers is not in the document]
Comment (yuki-stripe) [reply StripePayments/StripePayments/Source/PaymentHandler/STPPaymentHandler.swift:1023]: Ah thanks for pointing this out, this should actually be two separate guards.  If we're missing a currentAction, it should send an analytic. If we don't have a nextAction, we should fail the currentAction.

File: StripePayments/StripePayments/Source/PaymentHandler/STPPaymentHandler.swift [hunk 3]
```diff
@@ -998,45 +1008,51 @@ public class STPPaymentHandler: NSObject {
...
+            analyticsClient.log(analytic: errorAnalytic, apiClient: apiClient)
             return
         }
 
+        let failCurrentActionWithMissingNextActionDetails = {
+            currentAction.complete(
+                with: STPPaymentHandlerActionStatus.failed,
+                error: self._error(
+                    for: .unexpectedErrorCode,
```
[the remark this answers is not in the document]
Comment (yuki-stripe) [reply StripePayments/StripePayments/Source/PaymentHandler/STPPaymentHandler.swift:1032]: Yes, this particular case is an unexpected error - the next action type is supported, but somehow the next action details are missing.

File: StripePayments/StripePayments/Source/PaymentHandler/STPPaymentHandler.swift [hunk 4]
```diff
@@ -998,45 +1008,51 @@ public class STPPaymentHandler: NSObject {
             action.complete(with: STPPaymentHandlerActionStatus.canceled, error: nil)
 
         case .requiresSource:
-            fatalError()
+            action.complete(with: .failed, error: _error(for: .unexpectedErrorCode, loggingSafeUserInfo: ["error_message": "PaymentIntent status is requiresSource"]))
```
[the remark this answers is not in the document]
Comment (yuki-stripe) [reply StripePayments/StripePayments/Source/PaymentHandler/STPPaymentHandler.swift:1011]: Ooh this gives me an idea. Looks like we always call `error(for:loggingSafeUserInfo:)` with at most a single string that's basically an error message.  So I'll just refactor the method to be `func error(for:errorMessage:)`.

File: StripePayments/StripePayments/Source/PaymentHandler/STPPaymentHandler.swift [hunk 5]
```diff
@@ -1545,65 +1471,62 @@ public class STPPaymentHandler: NSObject {
...
-                                }
-                                let requiresAction: Bool = self._handlePaymentIntentStatus(
-                                    forAction: currentAction
+                            return
+                        }
+                        // If the transaction is still unexpectedly processing, refresh the PaymentIntent
+                        // This could happen if, for example, a payment is approved in an SFSafariViewController, the user closes the sheet, and the approval races with this fetch.
+                        if
+                            !STPPaymentHandler._isProcessingIntentSuccess(for: paymentMethod.type),
+                            retrievedPaymentIntent?.status == .processing && retryCount > 0
```
[the remark this answers is not in the document]
Comment (yuki-stripe) [reply StripePayments/StripePayments/Source/PaymentHandler/STPPaymentHandler.swift:1486]: TY!

File: StripePayments/StripePayments/Source/PaymentHandler/STPPaymentHandler.swift [hunk 6]
```diff
@@ -1183,250 +1133,188 @@ public class STPPaymentHandler: NSObject {
...
+                    currentAction.apiClient.authenticate3DS2(
+                        authParams,
+                        sourceIdentifier: useStripeSDK.threeDSSourceID ?? "",
+                        returnURL: currentAction.returnURLString,
+                        maxTimeout: currentAction.threeDSCustomizationSettings
+                            .authenticationTimeout,
+                        publishableKeyOverride: useStripeSDK.publishableKeyOverride
+                    ) { (authenticateResponse, error) in
+                        guard let authenticateResponse, error == nil else {
+                            let error = error ?? self._error(for: .stripe3DS2ErrorCode, loggingSafeErrorMessage: "Missing authenticate response")
```
Comment (wooj-stripe) [inline StripePayments/StripePayments/Source/PaymentHandler/STPPaymentHandler.swift:1156]: technically it is possible for us to enter this code if there is an `authenticationResponse` and there is an error, and we would log that we are missing an authentication response.  Maybe lets break this up a bit?

Also, might be more readable if we rename `authenticateResponse` to something like `result` or `authResult` or `authenticateResult`

File: StripePayments/StripePayments/Source/PaymentHandler/STPPaymentHandler.swift [hunk 7]
```diff
@@ -1183,250 +1133,188 @@ public class STPPaymentHandler: NSObject {
...
-                                                    finallyBlock: {
-                                                    }
+                                    guard self._canPresent(
+                                        with: currentAction.authenticationContext,
+                                        error: &presentationError
+                                    ) else {
+                                        currentAction.complete(
+                                            with: .failed,
+                                            error: presentationError
+                                        )
```
Comment (wooj-stripe) [inline StripePayments/StripePayments/Source/PaymentHandler/STPPaymentHandler.swift:1177]: Sorry, hard to follow the code, but when we complete an action with `.failed`, it seems like this eventually gets caught on line 407, w/ an assertionFailure?  I'm guessing we probably want to call stpAssert on this line?

File: StripePayments/StripePayments/Source/PaymentHandler/STPPaymentHandler.swift (hunk 6 above)
Comment (yuki-stripe) [reply StripePayments/StripePayments/Source/PaymentHandler/STPPaymentHandler.swift:1156]: That can't actually happen, we'll only use "Missing authenticate" if the error is nil. But it's very hard to read, so I'll break it up anyways.

I don't love the 'authenticateResponse' name but why would `result` etc be better? The class itself is called `STP3DS2AuthenticateResponse' and represents the `/authenticate` endpoint response.  Unless you feel strongly I'd prefer to put renaming out of scope for this PR.

File: StripePayments/StripePayments/Source/PaymentHandler/STPPaymentHandler.swift [hunk 8]
```diff
@@ -2095,13 +1990,16 @@ public class STPPaymentHandler: NSObject {
...
     @_spi(STP) public func _error(
         for errorCode: STPPaymentHandlerErrorCode,
         apiErrorCode: String? = nil,
-        loggingSafeUserInfo: [String: String] = [:]
+        loggingSafeErrorMessage: String? = nil,
+        localizedDescription: String? = nil
     ) -> NSError {
-        var userInfo = loggingSafeUserInfo
+        var userInfo = [String: String]()
+        userInfo[STPError.errorMessageKey] = loggingSafeErrorMessage
```
Comment (wooj-stripe) [inline StripePayments/StripePayments/Source/PaymentHandler/STPPaymentHandler.swift:2001]: Sorry, i didn't catch this before.. before we were calling:
```
_error( ...,
   loggingSafeUserInfo: ["error_message": "..."]
)
```

But now we end up calling:
```
_error(...,
      loggingSafeErrorMessage: "..."
)
//and then:
["com.stripe.lib:ErrorMessageKey": loggingSafeErrorMessage]
```
I'm not quite sure what key is expected here, but something seems off?

Comment (yuki-stripe) [reply StripePayments/StripePayments/Source/PaymentHandler/STPPaymentHandler.swift:2001]: Talked offline and resolved
